lib/python/numpy.h

```diff
@@ -157,36 +238,60 @@ bool memory_overlaps(const py::array_t<T> &data, const View &view) {
          std::greater<>()(data_end, view_begin);
 }
 
+/*
+ * The code here is not pretty.
+ * But a generic copy function would be much more complicated than the
+ * straightforward nested loops we use here.
+ * In practice, there is also little need to support ndim > 6 for non-contiguous
+ * data as transform does not support such variables either.
+ *
+ * For a working, generic implementation, see git ref
+ *  bd2e5f0a84d02bd5baf6d0afc32a2ab66dc09e2b
+ * and its history, in particular
+ *  86761b1e280a63b4f0b723a165188d21dd097972
+ *  8721b2d02b98c1acae5c786ffda88055551d832b
+ *  4c03a553827f2881672ae1f00f43ae06e879452c
+ *  c2a1e3898467083bf7d019a3cb54702c8b50ba86
+ *  c2a1e3898467083bf7d019a3cb54702c8b50ba86
+ */
 /// Copy all elements from src into dst.
 /// Performs an explicit conversion of elements in `src` to the element type of
-/// `dst` `convert == true`.
+/// `dst` if `convert == true`.
 /// Otherwise, elements in src are simply assigned to dst.
-template <bool convert, class T, class View>
-void copy_flattened(const py::array_t<T> &src, View &&dst) {
+template <bool convert, class T, class Dst>
+void copy_elements(const py::array_t<T> &src, Dst &dst) {
   if (scipp::size(dst) != src.size())
     throw std::runtime_error(
         "Numpy data size does not match size of target object.");
 
-  const auto dispatch = [](const py::array_t<T> &src_, View &&dst_) {
+  const auto dispatch = [&dst](const py::array_t<T> &src_) {
+    if (is_c_contiguous(src_))
+      return copy_flattened<convert>(src_, dst);
+
     switch (src_.ndim()) {
     case 0:
-      return copy_flattened_0d<convert>(src_, std::forward<View>(dst_));
+      return copy_flattened_0d<convert>(src_, dst);
     case 1:
-      return copy_flattened_1d<convert>(src_, std::forward<View>(dst_));
+      return copy_flattened_1d<convert>(src_, dst);
     case 2:
-      return copy_flattened_2d<convert>(src_, std::forward<View>(dst_));
+      return copy_flattened_2d<convert>(src_, dst);
     case 3:
-      return copy_flattened_3d<convert>(src_, std::forward<View>(dst_));
+      return copy_flattened_3d<convert>(src_, dst);
     case 4:
-      return copy_flattened_4d<convert>(src_, std::forward<View>(dst_));
+      return copy_flattened_4d<convert>(src_, dst);
+    case 5:
+      return copy_flattened_5d<convert>(src_, dst);
+    case 6:
+      return copy_flattened_6d<convert>(src_, dst);
     default:
       throw std::runtime_error("Numpy array has more dimensions than supported "
                                "in the current implementation.");
```

Comment:
Do you want to suggest in the error message to make a copy to a c-contiguous array as a workaround?